launchpad/launchpad.gno

```diff
@@ -0,0 +1,377 @@
+	"gno.land/r/gnoswap/v1/consts"
+	en "gno.land/r/gnoswap/v1/emission"
+	"gno.land/r/gnoswap/v1/gns"
+)
+
+var (
+	projects = make(map[string]Project) // projectId -> project
+
+	// project tier should distrubte project token if deposit ever happend
+	// therefore we need to keep track of project tiers without deposit
+	projectTiersWithoutDeposit = make(map[string]bool) // tierId -> true
+)
+
+var (
+	q96 = u256.Zero()
+)
+
+const (
+	PAD_SEP = "*PAD*"
+)
+
+func init() {
+	q96 = u256.MustFromDecimal(consts.Q96)
+}
+
+// ProjectInput represents the input parameters for creating a new project
+type ProjectInput struct {
+	Name             string
+	TokenPath        string
+	Recipient        std.Address
+	DepositAmount    uint64
+	ConditionsToken  string
+	ConditionsAmount string
+	Tier30Ratio     uint64
+	Tier90Ratio     uint64
+	Tier180Ratio    uint64
+	StartTime       uint64
+}
+
+// validateProjectInput validates the input parameters for creating a new project
+func validateProjectInput(input ProjectInput, now uint64) error {
+	if input.Tier30Ratio+input.Tier90Ratio+input.Tier180Ratio != 100 {
+		return errors.New(ufmt.Sprintf("invalid ratio, sum of all tiers(30:%d, 90:%d, 180:%d) should be 100", 
+			input.Tier30Ratio, input.Tier90Ratio, input.Tier180Ratio))
+	}
+
+	if input.StartTime <= now {
+		return errors.New(ufmt.Sprintf("invalid start time(%d), cannot start project in past(now:%d)", 
+			input.StartTime, now))
+	}
+
+	if input.DepositAmount == 0 {
+		return errors.New("cannot deposit 0")
+	}
+
+	if !input.Recipient.IsValid() {
+		return errors.New(ufmt.Sprintf("invalid recipient address(%s)", input.Recipient.String()))
+	}
+
+	return nil
+}
+
+// validateToken checks if the token is registered
+func validateToken(tokenPath string, registeredTokens []string) error {
+	for _, regToken := range registeredTokens {
+		if regToken == tokenPath {
+			return nil
+		}
+	}
+	return errors.New(ufmt.Sprintf("token(%s) not registered", tokenPath))
+}
+
+// createTier creates a new tier with the given parameters
+func createTier(projectId string, duration uint64, amount uint64, startHeight uint64, 
+	startTime uint64, collectWaitDuration uint64) Tier {
+	
+	tierId := generateTierId(projectId, duration)
+	endTime := startTime + (duration * TIMESTAMP_DAY)
+	endHeight := startHeight + (duration * TIMESTAMP_DAY * 1000 / uint64(gns.GetAvgBlockTimeInMs()))
+
+	tier := Tier{
+		id:                  tierId,
+		collectWaitDuration: collectWaitDuration,
+		tierAmount:          amount,
+		ended: TimeInfo{
+			height: endHeight,
+			time:   endTime,
+		},
+	}
+
+	projectTiersWithoutDeposit[tierId] = true
+	return tier
+}
+
+func CreateProject(
+	name string,
+	tokenPath string,
+	recipient std.Address,
+	depositAmount uint64,
+	conditionsToken string,
+	conditionsAmount string,
+	tier30Ratio uint64,
+	tier90Ratio uint64,
+	tier180Ratio uint64,
+	startTime uint64,
+) string {
+	common.IsHalted()
+
+	if err := common.SatisfyCond(isUserCall()); err != nil {
```

Comment:
doesn't matter if user or contract called this function.
AdminOnly will do the jobs

Reply:
https://github.com/gnoswap-labs/gnoswap/pull/451/commits/f5b33c63e8245d54020a32e7c2bf171cbdf77e86